### PR TITLE
`assert_almost_eq!`

### PR DESCRIPTION
**Pull request by mo271**

### CODE REVIEW

File: jxl_test_utils/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+#[macro_export]
```
Comment (veluca93) [inline jxl_test_utils/src/lib.rs:6]: `pub use assert_almost_eq` works if staying in the same crate.

File: jxl_test_utils/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        }
+    };
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_with_floats() {
```
Comment (veluca93) [inline jxl_test_utils/src/lib.rs:33]: Can you also test NaNs?

**Review by veluca93 — COMMENTED** [review]
Isn't a new crate a bit overkill? Feels like a module in `util/test.rs` would be sufficient...

File: jxl_test_utils/src/lib.rs (hunk 2 above)
Comment (mo271) [reply jxl_test_utils/src/lib.rs]: done, also INFINITY

File: jxl_test_utils/src/lib.rs (hunk 1 above)
Comment (mo271) [reply jxl_test_utils/src/lib.rs]: move to util/test.rs

File: jxl/src/util/test.rs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        );
+
+        assert_almost_eq!(1.0, 1.0, INFINITY);
+
+        let result = panic::catch_unwind(|| {
+            assert_almost_eq!(NAN, NAN, INFINITY);
+        });
+        assert!(
+            result.is_err(),
+            "Expected assert_almost_eq! to panic, but it didn't"
```
Comment (veluca93) [inline jxl/src/util/test.rs:90]: https://doc.rust-lang.org/rust-by-example/testing/unit_testing.html#testing-panics :-)

File: jxl/src/util/test.rs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        );
+
+        let result = panic::catch_unwind(|| {
+            assert_almost_eq!(INFINITY, INFINITY, INFINITY);
```
Comment (veluca93) [inline jxl/src/util/test.rs:94]: It's surprising to me that this should panic.

Comment (veluca93) [reply jxl/src/util/test.rs]: I agree that this is what *happens*, but I am not sure I agree that this is what _should_ happen ;)

File: jxl/src/util.rs [hunk 5]
```diff
@@ -15,3 +15,5 @@ pub use bits::*;
 pub use concat_slice::*;
 pub use log2::*;
 pub use shift_right_ceil::*;
+
+mod test;
```
Comment (veluca93) [inline jxl/src/util.rs:19]: I assume you want this to be pub ;-)

File: jxl/src/util/test.rs (hunk 4 above)
Comment (veluca93) [reply jxl/src/util/test.rs]: Then again probably we don't want infinities to be around anyway, so this might be fine

File: jxl/src/util/test.rs [hunk 6]
```diff
@@ -0,0 +1,84 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+#[macro_export]
```
Comment (veluca93) [inline jxl/src/util/test.rs:6]: I think `pub use assert_almost_eq` and then importing `util::test::assert_almost_eq` would be better

File: jxl/src/util.rs (hunk 5 above)
Comment (mo271) [reply jxl/src/util.rs]: done

File: jxl/src/util/test.rs (hunk 6 above)
Comment (mo271) [reply jxl/src/util/test.rs]: using `pub(crate)...` now

File: jxl/src/render/stages/upsample.rs [hunk 7]
```diff
@@ -87,7 +87,11 @@ impl RenderPipelineStage for Upsample2x {
 #[cfg(test)]
 mod test {
     use super::*;
-    use crate::{error::Result, image::Image, render::test::make_and_run_simple_pipeline};
+    //use util::test::assert_almost_eq;
```
Comment (veluca93) [inline jxl/src/render/stages/upsample.rs:90]: leftover

Comment (mo271) [reply jxl/src/render/stages/upsample.rs]: removed...

**Pull request merged**